Extensions/Firebase/C_firebasetools/D_storagetools.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Firebase Tools Collection
+ * @fileoverview
+ * @author arthuro555
+ */
+
+/**
+ * Firebase Storage Event Tools
+ * @namespace
+ */
+gdjs.evtTools.firebase.storage = {
+  uploads: new gdjs.UIDArray(),
+};
+
+/**
+ * Uploads a file as string to the firebase storage bucket.
+ * @param {string} file - The entire file as string.
+ * @param {string} onlinePath - The path under wich the file will be accessible on the bucket.
+ * @param {"none"|"base64"|"base64url"|"data_url"} [type] - The type/format of the string to upload.
+ * @param {gdjs.Variable} [callbackValueVariable] - The variable where to store the result (url to the file).
+ * @param {gdjs.Variable} [callbackStateVariable] - The variable where to store if the operation was successful.
+ * @param {gdjs.Variable} [callbackUIDVariable] - The variable where to store the upload ID.
+ * @param {gdjs.Variable} [callbackProgressVariable] - The variable where to store the progress.
+ */
+gdjs.evtTools.firebase.storage.upload = function (
+  file,
+  onlinePath,
+  type,
+  callbackValueVariable,
+  callbackStateVariable,
+  callbackUIDVariable,
+  callbackProgressVariable
+) {
+  type = type === 'none' ? undefined : type;
+
+  let uploadTask;
+  try {
+    uploadTask = firebase
+      .storage()
+      .ref()
+      .child(onlinePath)
+      .putString(file, type);
+  } catch (e) {
+    if (typeof callbackStateVariable !== 'undefined')
+      callbackStateVariable.setString(e.message);
+    return;
+  }
+
+  let uploadID;
+  if (typeof callbackUIDVariable !== 'undefined') {
+    uploadID = gdjs.evtTools.firebase.storage.uploads.push(uploadTask); // Only bother pushing if the ID will be stored
+    if (typeof callbackUIDVariable !== 'undefined')
+      callbackUIDVariable.setNumber(uploadID);
+  }
+
+  uploadTask.on(
+    firebase.storage.TaskEvent.STATE_CHANGED,
+    (uploadProgress) => {
+      if (typeof callbackProgressVariable !== 'undefined')
+        gdjs.evtTools.network._objectToVariable(
+          uploadProgress,
+          callbackProgressVariable,
+          ['tasks'] // Remove circular reference
+        );
+    },
+    (error) => {
+      if (typeof callbackStateVariable !== 'undefined')
+        callbackStateVariable.setString(error.message);
+    },
+    () => {
+      if (typeof callbackStateVariable !== 'undefined')
+        callbackStateVariable.setString('ok');
+      if (typeof callbackUIDVariable !== 'undefined') {
+        gdjs.evtTools.firebase.storage.uploads.remove(uploadID); // Free memory
```

Comment:
Can you explain why this is in the `if`?

Reply:
In case a user tries to use it via JS, it will crash if they don't add callbacks, and they migh not pass them if they don't need them.

Reply:
The question is specific about this line: we're testing for `callbackUIDVariable` but then using `gdjs.evtTools.firebase.storage.uploads.remove(uploadID)`, which seems unrelated?

Reply:
If the user doesn't pass a variable, a uid is not generated (as the user won't store it anyways) so if there is no Variable, there is no UID to remove